smdebug_rulesconfig/actions/actions.py

```diff
@@ -0,0 +1,177 @@
+from smdebug_rulesconfig.actions.utils import validate_string
+
+
+class Action(object):
+    def __init__(self, **action_parameters):
+        """
+        Base class for action, which is to be invoked when a rule fires.. Offers `serialize` function to convert action
+        parameters to a string dictionary. This class is not meant to be initialized directly. Accepts dictionary of
+        action parameters and drops keys whose values are `None`.
+
+        :param action_parameters: Dictionary of action parameters.
+        """
+        action_parameters["name"] = self.__class__.__name__.lower()
+        self.action_parameters = {
+            key: value for key, value in action_parameters.items() if value is not None
+        }
+
+    def serialize(self):
+        """
+        Serialize the action parameters as a string dictionary.
+
+        :return: Action parameters serialized as a string dictionary.
+        """
+        return (
+            "{"
+            + ", ".join(
+                [f'\\"{key}\\": \\"{value}\\"' for key, value in self.action_parameters.items()]
+            )
+            + "}"
+        )
```

Comment:
you should validate this json str before returning using your validate method and raise appropriate exception message if it is invalid.

Same with other serialize methods.

Reply:
The arguments (`self.action_parameters`) are validated by this point, and the json str is formed by us. We have tests that validate this json str. User should not be facing exception with json str.

Reply:
playing devils advocate here- What if i provide a string like vikas\"\@gmail.com (some escaped string which makes your json creation invalid)? 
validating string means the json string that we created should be valid, otherwise it is going to fail later in rules, which we want to avoid as early as possible. We can avoid that by makign sure that the json that we created out of parameters are indeed valid, if not we raise appropriate errors in SDK itself.

Reply:
Makes sense. I think this falls under the argument validation though, so I'll update that to be more strict.

Reply:
Should you call the same function, that you wrote in test/core/utils in this PR? 
 Ultimately you want is the end string that serialize is returning, is valid as seen by rules container. 
 Also key thing would be how do we raise proper error mesg for customer, so that mesg clearly specifies how can it be corrected.

Reply:
We should have strict validation on any user arguments so that ideally we don't need to rely on that function to catch errors from the user. Regarding that function I wrote, we would need to raise a proper error message for customer as you mentioned. And we can't, as I mentioned in a separate comment thread.

So the best solution is to use strict argument validation (regex) to catch all potential errors. Perhaps we should still call that function anyway since user experience is better if error is caught early on even if there is no helpful message.

Reply:
>So the best solution is to use strict argument validation (regex) to catch all potential errors. Perhaps we should still call that function anyway since user experience is better if error is caught early on even if there is no helpful message.

Yeah i agree. 
so, 
validate_params()
create_json
validate json

Reply:
Done.